### PR TITLE
Add debugging option to write vmservice address to file after starting

### LINKED ISSUE #33129: Add ability to run `flutter driver` tests on multiple devices at the same time
## Use case

As a developer, I would like to be able to run the same suite of integration tests on all the devices connected to my machine. This is will allow me to test my application in many different devices in parallel, being able to obtain results faster.

## Proposal

There could be two different proposals:

- Allow multiple instances of `flutter drive` to run at the same time, so I could run the following commands without getting the error `Waiting for another flutter command to release the startup lock...`
    - Terminal1: `flutter drive ... -d iphone`
    - Termina2: `flutter drive ... -d pixel`
- Allow the `-d` flag (devices) to pass the value `all`, running integration tests in all the devices connected to that computer: `flutter drive ... -d all`

### PR DESCRIPTION
**Pull request by jonahwilliams**
## Description

An attempt at addressing the workflow in https://github.com/flutter/flutter/issues/33129. Adds a new option to run, `--vmservice-out-file`. This takes a String which is interpreted as a path to a file where the URI of the _first_ connected vmservice is written. The intention is that this should make automation of flutter build/run/driver easier.

### Example workflow

The example below scripts `flutter run` and `flutter drive` to start and launch a driver test:

```shell
# Run the application
flutter run -d deviceId \
  --target=test_driver/example_main.dart \ # ensure we use the test driver entrypoint
  --start-paused \ # keep the vm paused until the driver starts
  --no-resident \ # disconnect but keep the application running
  --vmservice-out-file=foo  # NEW! write the vmservice uri to the file foo for reuse. 

# The above command will exit once it is finished, and we can start the flutter driver:
flutter drive \
  --use-existing-app="$(cat foo)" \ # NEW! grab the vmservice uri we wrote above
  --driver=test_driver/example_main_test.dart  # provide the test script
```

Fixes #33129

### CODE REVIEW

File: packages/flutter_tools/lib/src/run_cold.dart [hunk 1]
```diff
@@ -114,6 +114,12 @@ class ColdRunner extends ResidentRunner {
 
     appStartedCompleter?.complete();
 
+    if (debuggingOptions.vmserviceOutFile != null) {
+      final File vmserviceOutFile = fs.file(debuggingOptions.vmserviceOutFile);
+      vmserviceOutFile.createSync(recursive: true);
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/run_cold.dart]: Should we catch exceptions if creating the file fails?

File: packages/flutter_tools/lib/src/run_hot.dart [hunk 2]
```diff
@@ -229,6 +229,12 @@ class HotRunner extends ResidentRunner {
       return 0;
     }
 
+    if (debuggingOptions.vmserviceOutFile != null) {
+      final File vmserviceOutFile = fs.file(debuggingOptions.vmserviceOutFile);
+      vmserviceOutFile.createSync(recursive: true);
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/run_hot.dart]: same question here

File: packages/flutter_tools/lib/src/run_cold.dart (hunk 1 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/run_cold.dart]: Yes!

File: packages/flutter_tools/lib/src/run_cold.dart [hunk 3]
```diff
@@ -114,6 +114,12 @@ class ColdRunner extends ResidentRunner {
...
+      vmserviceOutFile.writeAsStringSync(flutterDevices.first.vmServices.first.wsAddress.toString());
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/run_cold.dart]: Is picking the first vmService arbitrary (not sure why a device would have more than one)?

Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/run_cold.dart]: We don't support any debugging functionality for multiple devices, so by convention we always return the first vmservice address

File: packages/flutter_tools/lib/src/run_cold.dart (hunk 1 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/run_cold.dart]: Fixed by moving into base class, adding a catch and a new test case

File: packages/flutter_tools/lib/src/run_hot.dart (hunk 2 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/run_hot.dart]: Done with above

**Comment by HerrNiklasRaab** [comment]
@jonahwilliams When I run flutter with the `--no-resident` the vmservice url written to the file is closed after the command has completed. 

This is the command that I am executing:
`flutter run -t test_driver/app.dart --observatory-port=8888 --disable-service-auth-codes --flavor=development --no-resident --vmservice-out-file=foo --start-paused`

While running the observatory, becomes available for a short time, but after the command has completed, it's closed. Do you know why?

**Comment by HerrNiklasRaab** [comment]
When you run the first command without `no-resident`, then wait until the URL is printed to the file. And then start the second command with the URL from the file, in a separate terminal than it works.

**Pull request merged**